### PR TITLE
Future prediction

### PR DESCRIPTION
**Pull request by lukeeffenberger**
Hoffe ist jetzt alles richtig so!

### CODE REVIEW

**Comment by pnieters** [comment]
Can you give a quick point per point which files have changed and why?
-- will anything that is currently in-use break?

**Comment by lukeeffenberger** [comment]
> Can you give a quick point per point which files have changed and why?
> -- will anything that is currently in-use break?

I merged master into it and let the script run. nothing broke. yes will add some comments

File: src/BaseModel.py [hunk 1]
```diff
@@ -114,7 +114,7 @@ def call(self, t, x):
 class IAEffectLoader(object):
     generates_stats = False
 
-    def __init__(self, var, filenames, days, counties):
+    def __init__(self, var, filenames, days, counties, predict_for=None):
```
Comment (lukeeffenberger) [inline src/BaseModel.py:117]: ia effect loader needs to repeat the last iaeffects if we want to predict into the future

File: src/BaseModel.py [hunk 2]
```diff
@@ -132,6 +132,19 @@ def __init__(self, var, filenames, days, counties):
...
+                    d1 = [ds.index(d) for d in days]
+                    d2 = list(range(d1[-1],d1[-1]+len(predict_for)))
+                    n_days_pred = len(d2)
+                    # Repeat ia_effects for last day.
+                    last = m[-1,:,:]
+                    last = np.tile(last,(n_days_pred,1,1))
+                    m = np.concatenate((m,last), axis=0)
+                    # Update d_idx.
+                    d_idx = np.array(d1 + d2).reshape(-1,1)
+
```
Comment (lukeeffenberger) [inline src/BaseModel.py:147]: repeat the ia effects

File: src/BaseModel.py [hunk 3]
```diff
@@ -397,16 +410,26 @@ def sample_predictions(
             target_days,
             target_counties,
             parameters,
+            prediction_days,
```
Comment (lukeeffenberger) [inline src/BaseModel.py:413]: provide future days for which to predict

File: src/BaseModel.py [hunk 4]
```diff
@@ -397,16 +410,26 @@ def sample_predictions(
             target_days,
             target_counties,
             parameters,
+            prediction_days,
+            average_periodic_feature=True,
```
Comment (lukeeffenberger) [inline src/BaseModel.py:414]: whether or not the periodic feature is to be averaged

File: src/BaseModel.py [hunk 5]
```diff
@@ -397,16 +410,26 @@ def sample_predictions(
...
             target_counties,
             parameters,
+            prediction_days,
+            average_periodic_feature=True,
             init="auto"):
+        all_days = pd.DatetimeIndex([d for d in target_days] + [d for d in prediction_days])
+        
         # extract features
-        features = self.evaluate_features(target_days, target_counties)
+        features = self.evaluate_features(all_days, target_counties)
```
Comment (lukeeffenberger) [inline src/BaseModel.py:419]: evaluate the feature for the training and for the prediction days

File: src/gridjob_sample_posterior_predictions.sge [hunk 6]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
```
Comment (lukeeffenberger) [inline src/gridjob_sample_posterior_predictions.sge:1]: whole script probably irrelevant, only my grid script

File: src/plot_curves_appendix.py [hunk 7]
```diff
@@ -1,4 +1,7 @@
+import matplotlib
+matplotlib.use('TkAgg')
```
Comment (lukeeffenberger) [inline src/plot_curves_appendix.py:2]: this line is needed when working on Mac, but didn't break on linux I think

File: src/plot_curves_appendix.py [hunk 8]
```diff
@@ -79,46 +82,51 @@ def curves_appendix(use_interactions=True, use_report_delay=True, save_plot=Fals
     disease = "covid19"
     prediction_region = "germany"
 
-    data = load_daily_data(disease, prediction_region, county_info)
+
+    days_into_future = 5
+    data = load_daily_data(disease, prediction_region, counties, pad=days_into_future)
```
Comment (lukeeffenberger) [inline src/plot_curves_appendix.py:87]: when loading the data the csv file has to be extended/padded with empty lines

File: src/plot_curves_appendix.py [hunk 9]
```diff
@@ -79,46 +82,51 @@ def curves_appendix(use_interactions=True, use_report_delay=True, save_plot=Fals
...
     # Load our prediction samples
     res = load_pred(disease, use_interactions, use_report_delay)
-    n_days = ((last_day - pd.Timedelta(days=1)) - first_day).days
+    n_days = (last_day  - first_day).days
+    print(n_days)
 
     prediction_samples = np.reshape(res['y'], (res['y'].shape[0], n_days, -1)) 
     prediction_quantiles = quantiles(prediction_samples, (5, 25, 75, 95))
 
+    ext_index = pd.DatetimeIndex([d for d in target.index] + [d for d in pd.date_range(target.index[-1]+timedelta(1), last_day-timedelta(1))])
```
Comment (lukeeffenberger) [inline src/plot_curves_appendix.py:107]: index for training days + prediction days

File: src/plot_curves_appendix.py [hunk 10]
```diff
@@ -166,7 +174,7 @@ def curves_appendix(use_interactions=True, use_report_delay=True, save_plot=Fals
             zorder=3)
 
         # plot ground truth
-        p_real = ax.plot(days, target[county_id], "k.")
+        p_real = ax.plot(days[:-4], target[county_id], "k.")
```
Comment (lukeeffenberger) [inline src/plot_curves_appendix.py:177]: plot ground truth only where data Is available

File: src/sample_posterior.py [hunk 11]
```diff
@@ -58,24 +65,33 @@
...
+        target_train,
+        samples=num_samples,
+        tune=100,
+        target_accept=0.95,
+        max_treedepth=15,
+        chains=num_chains,
+        cores=num_cores)
+
+    with open(filename_model, "wb") as f:
+    	pkl.dump(model.model, f)
```
Comment (lukeeffenberger) [inline src/sample_posterior.py:80]: added possibility to load parameters if they were already sampled

File: src/shared_utils.py [hunk 12]
```diff
@@ -70,8 +71,16 @@ def load_daily_data(disease, prediction_region, counties, seperator=","):
...
 
+    if pad is not None:
+        # get last date
+        last_date = pd.Timestamp(data.iloc[:, -1].index[-1])
+        extra_range = pd.date_range(
+            last_date+timedelta(1), last_date+timedelta(pad))
+        for x in extra_range:
+            data = data.append(pd.Series(name=str(x)[:11]))
+
+    data.index = [pd.Timestamp(date) for date in data.index]
```
Comment (lukeeffenberger) [inline src/shared_utils.py:83]: padding the data

**Review by lukeeffenberger — COMMENTED** [review]
kannst du so meine Kommentare sehen?

**Review by pnieters — APPROVED** [review]

**Pull request merged**